### PR TITLE
Add message-format=json support, second take

### PR DESCRIPTION
**Pull request by konstin**
This is #53 with the merge conflicts resolved and the review comments addressed. I also copied `DiagnosticLevel` from @epage.

CC @roblabla

### CODE REVIEW

**Comment by roblabla** [comment]
Thanks for doing this. I've been meaning to do it but haven't been able to find the time 😓 .

The vast majority of messages from cargo will be of the FromCompiler, so I don't think forcing one more allocation in that path is a good idea, especially when the gains are fairly trivial. I feel like the warning should be ignored in this case.

File: src/diagnostic.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    pub message: String,
+    /// The associated error code for this diagnostic
+    pub code: Option<DiagnosticCode>,
+    /// "error: internal compiler error", "error", "warning", "note", "help"
+    pub level: DiagnosticLevel,
+    /// A list of source code spans this diagnostic is associated with.
+    pub spans: Vec<DiagnosticSpan>,
+    /// Associated diagnostic messages.
+    pub children: Vec<Diagnostic>,
+    /// The message as rustc would render it
```
[the remark this answers is not in the document]
Comment (konstin) [reply src/diagnostic.rs]: Can we just assume that everyone uses a reasonably recent compiler version?

**Comment by konstin** [comment]
Thanks for the review @ehuss, I've fixed most doc strings now.

File: src/dependency.rs [hunk 2]
```diff
@@ -17,7 +17,8 @@ pub enum DependencyKind {
     /// Those used in build scripts only
     Build,
     #[doc(hidden)]
-    DoNotMatchExhaustively,
+    #[serde(other)]
+    Unknown,
```
[the remark this answers is not in the document]
Comment (konstin) [reply src/dependency.rs]: No, iirc #[serde(other)] is just the saner version of DoNotMatchExhaustively (https://github.com/oli-obk/cargo_metadata/pull/53#issuecomment-419975385)

**Comment by konstin** [comment]
I'd love to have a base crate shared with cargo. Imo we should track that in one central issue so information doesn't get lost across thread, maybe directly in the cargo repo.

Personally, I'd like to have some kind of `cargo_shared_types` crate that contains types for all the json cargo emits, which is currently diagnostics, `cargo metadata` and the build plan. But for now I'd like to have this merged and released before doing bigger refactorings work because I'm already using this branch in pyo3-pack.

**Comment by oli-obk** [comment]
> The cargo team would like to have the cargo_diagnostics crate under their organization. I think it's fine to start out in a private repo and let them fork or move it

Actually they just want it to be a subcrate of the cargo repo. Can you make a crate out of each of the two modules and add them to the cargo repo?

**Pull request merged**